temporalio/envconfig.py

```diff
@@ -0,0 +1,408 @@
+            client_private_key=_read_source(self.client_private_key),
+        )
+
+    @staticmethod
+    def _from_dict(d: Optional[Mapping[str, Any]]) -> Optional[ClientConfigTLS]:
+        if not d:
+            return None
+        return ClientConfigTLS(
+            disabled=d.get("disabled", False),
+            server_name=d.get("server_name"),
+            # Note: Bridge uses snake_case, but TOML uses kebab-case which is
+            # converted to snake_case. Core has server_ca_cert, client_key.
+            server_root_ca_cert=_from_dict_to_source(d.get("server_ca_cert")),
+            client_cert=_from_dict_to_source(d.get("client_cert")),
+            client_private_key=_from_dict_to_source(d.get("client_key")),
+        )
+
+
+class ClientConnectConfig(TypedDict, total=False):
+    """Arguments for `temporalio.client.Client.connect` that are configurable via
+    environment configuration.
+
+    .. warning::
+        Experimental API.
+    """
+
+    target_host: Optional[str]
+    namespace: Optional[str]
+    api_key: Optional[str]
+    tls: Optional[Union[bool, temporalio.service.TLSConfig]]
+    rpc_metadata: Optional[Mapping[str, str]]
+
+
+@dataclass(frozen=True)
+class ClientConfigProfile:
+    """Represents a client configuration profile.
+
+    This class holds the configuration as loaded from a file or environment.
+    See `to_connect_config` to transform the profile to `ClientConnectConfig`,
+    which can be used to create a client.
+
+    .. warning::
+        Experimental API.
+    """
+
+    address: Optional[str] = None
+    """Client address."""
+    namespace: Optional[str] = None
+    """Client namespace."""
+    api_key: Optional[str] = None
+    """Client API key."""
+    tls: Optional[ClientConfigTLS] = None
+    """TLS configuration."""
+    grpc_meta: Mapping[str, str] = field(default_factory=dict)
+    """gRPC metadata."""
+
+    @staticmethod
+    def from_dict(d: Mapping[str, Any]) -> ClientConfigProfile:
+        """Create a ClientConfigProfile from a dictionary."""
+        return ClientConfigProfile(
+            address=d.get("address"),
+            namespace=d.get("namespace"),
+            api_key=d.get("api_key"),
+            tls=ClientConfigTLS._from_dict(d.get("tls")),
+            grpc_meta=d.get("grpc_meta") or {},
+        )
+
+    def to_client_connect_config(self) -> ClientConnectConfig:
+        """Create a `ClientConnectConfig` from this profile."""
+        config: ClientConnectConfig = {}
+        if self.address:
+            config["target_host"] = self.address
+        if self.namespace:
+            config["namespace"] = self.namespace
+        if self.api_key:
+            config["api_key"] = self.api_key
+        if self.tls:
+            config["tls"] = self.tls.to_connect_tls_config()
+        if self.grpc_meta:
+            config["rpc_metadata"] = self.grpc_meta
+        return config
+
+
+@dataclass
+class ClientConfig:
```

Comment:
We want users to be able to convert this (and `ClientConfigProfile`) to/from TOML on their own if they want. However, we don't necessarily need to expose literal TOML conversion, just the ability to convert to/from the `dict` that represents TOML (they can use some Python TOML library if they want to do the TOML part). Can we expose `to_dict` and `from_dict` on this and the `ClientConfigProfile` class that represents exactly what it looks like in TOML?

Reply:
added `to_dict` methods
the resulting dictionaries are well-typed, not sure if that was necessary or not